xarray/backends/pnc_.py

```diff
@@ -0,0 +1,91 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import functools
+
+import numpy as np
+
+from .. import Variable
+from ..core.utils import (FrozenOrderedDict, Frozen)
+from ..core import indexing
+
+from .common import AbstractDataStore, DataStorePickleMixin, BackendArray
+
+
+class PncArrayWrapper(BackendArray):
+
+    def __init__(self, variable_name, datastore):
+        self.datastore = datastore
+        self.variable_name = variable_name
+        array = self.get_array()
+        self.shape = array.shape
+        self.dtype = np.dtype(array.dtype)
+
+    def get_array(self):
+        self.datastore.assert_open()
+        return self.datastore.ds.variables[self.variable_name]
+
+    def __getitem__(self, key):
+        key = indexing.unwrap_explicit_indexer(
+            key, target=self, allow=indexing.BasicIndexer)
+
+        with self.datastore.ensure_open(autoclose=True):
+            array = self.get_array()
+            if key == () and self.ndim == 0:
+                return array[...]
+            return array[key]
+
+
+class PncDataStore(AbstractDataStore, DataStorePickleMixin):
+    """Store for accessing datasets via PseudoNetCDF
+    """
+
+    def __init__(self, filename, mode='r', autoclose=False):
+        from PseudoNetCDF import pncopen
+        try:
+            opener = functools.partial(pncopen, filename,
```

Comment:
For new backend classes, we're trying to move to a style where there is a dedicated classmethod `open` that handles opening a file from scratch and the backend constructor accepts an already open File object. Take a look at NetCDF4Backend for an example.

Reply:
This would require changes to many of my classes that accept paths. Is this a preconditions?

Reply:
Can you elaborate a little bit -- how do you normally call `pncopen`? I assume it accepts some set of required and optional arguments?

For the usual case, users would write:
```
# this would call pncopen(filename) internally
ds = xr.open_dataset(path, engine='pseudonetcdf')
```
For advanced cases, users would write something like:
```
# this would call pncopen(filename, mode='r', format='netcdf') internally
store = xr.backends.PsuedoNetCDFBackend.open(filename, mode='r', format='netcdf')
ds = xr.open_dataset(store)
```

Reply:
A normal call looks like this

```
myfile = pncopen(path, format=fmt, **kwds)
```

The only required argument is path. The format keyword is a named format and is highly recommended. If format is omitted, pncopen will cycle through all the formats it knows using the isMine class method to find a class that can open it. The other keywords are optional from pncopen's perspective, but may be required if one is opening file formats that are not fully self-describing.

I had not realized that you can provide the `open` interface and the resulting store to open_dataset. I was not planning to provide support for these because I assume your __init__ interfaces need to be identical. This seems like a nice alternative.

I'll make an adjustment here.

Reply:
We should probably add an optional `backend_kwargs` argument to `open_dataset` to make this a little easier.